### ISSUE TITLE
Validity of requirements file

### ISSUE THREAD

**Issue by CBenghi** [buildingSMART/IDS#89]
Hello,

Since it's possible to have invalid configurations of `Specifications`, `Facets` and `SimpleValues`, I suggest that we should standardise the rules that define an invalid specification.

My thinking is that we want to avoid a situation where a model fails the verification due to an invalid requirement.

I'm reporting here some considerations that we have started in #85:

Invalid elements might be:

- facets have required properties (e.g. Entity requires Name), if such properties are missing, then a facet is invalid
- a pattern constraint with wrong pattern syntax would invalidate the parent facet
- a maxOccur smaller than a minOccur

Maybe some more refined rules could be conceived such as:

- using an inverse attribute name in the attribute facet (I read they are considered invalid)
- using a non existing class name in the entity facet

And even more complex such as:

- using a regex pattern in attribute name that does not match any existing ifc attribute

If a facet is invalid would invalidate the parent requirement, and these would not be used when verifying a model.

**Comment by Moult** [comment]
Note that the docs already include implementation guidance:

 - https://github.com/buildingSMART/IDS/blob/master/Documentation/developer-guide.md#authoring-ids
 - https://github.com/buildingSMART/IDS/blob/master/Documentation/developer-guide.md#checking-ids-against-ifc

As for your points:

 - facets have required properties - this is validated by the XSD, no need for further docs
 - pattern constraint that excludes another facet - it's valid, but the user being silly. I don't see the need for this to be a strict rule. Maybe we can build a "helper" that checks for these things.
 - a maxoccur smaller than minoccurs - the docs already spell out the valid combos of occurs: https://github.com/buildingSMART/IDS/blob/master/Documentation/developer-guide.md#optionality
 - using inverse attributes - already as part of this guide: https://github.com/buildingSMART/IDS/blob/master/Documentation/developer-guide.md#authoring-ids
 - using non-existing class names - already as part of this guide: https://github.com/buildingSMART/IDS/blob/master/Documentation/developer-guide.md#authoring-ids
 - using a regex pattern in attribute names - if the dev is motivated, it's already part of this guide: https://github.com/buildingSMART/IDS/blob/master/Documentation/developer-guide.md#authoring-ids

I'm not sure if there is much that needs to be done. The docs state "An IDS is considered valid if it passes the XSD-based validation check.", and that's correct from a syntax perspective, and the rest are covered in testcases "Any software implementing IDS checking must comply with the following test suites." and implementer guides.

**Comment by CBenghi** [comment]
Hello @Moult,

I am ashamed to say that I left out the entire point I wanted to make (I had made it in another issue, and failed to report here).

I think it's misleading that we fail a model where the problem is in the specification (see for instance the use of inverses in attributes).
invalid requirements should not be run against a model because it could create confusion in the supply chain of real projects, misplacing the blame for poor IDS onto the party producing the model (e.g. what it's currently the case with inverse attributes).

I'm not suggesting this is a documentation problem, but a problem with limiting the outcomes of IDS to pass/fail, that we should address, and that implementers should include in their code.

To achieve this we should of course agree on what is valid/invalid, and since we are placing stricter restrictions on the IDS data than what is captured by the XSD schema, beyond the documentation, which I'm not disputing, have examples of VALID/INVALID requirements

I would be happy to work on this, if agreed, the list of points that I made is only to start thinking of possible cases that should be checked.

**Comment by Moult** [comment]
I understand now. I see two possible solutions:

Option 1 is to specify the invalid checks to be done and any IDS tool must comply with it. Potentially offering a third state in addition to pass and fail... An invalid state.

Option 2 is for building smart to provide an IDS validator. This'll do a two part validation, one doing an XSD check and second doing a schema check and so on.

Option 2 is my preference, and it should be free and open source for anyone to use. If vendors want it "built in" to guide users, that's even better, but at least there's an official validator.

**Comment by aothms** [comment]
Mine too, +1 on opt 2 (and I think something was already commissioned by bsi). Making a third state normative is going to blow up the implementation effort for the simpler toolboxes that don't have extensive schema knowledge.

**Comment by CBenghi** [comment]
This looks like a good solution to me.
I'd be happy to release that code under MIT licensing.

Would you agree that coherence of multiple implementations requires a shared set of test cases for valid/invalid?

**Comment by aothms** [comment]
The benefit of Option 2 though is that there isn't an immense time pressure to complete this, because the behaviour of invalid spec -> no match is still the correct behaviour. There will be some hand holding anyway in the early days and end-users will be busy creating trivial property rules. And then the set of ids-validation rules can also grow to include advise on formulating things better, "did you mean X?" kind of things.

In an ideal world I would still see them implemented in an implementation agnostic language such as xslt or schematron, but I already know that's probably not going to happen.

**Comment by CBenghi** [comment]
> the behaviour of invalid spec -> no match is still the correct behaviour. 

Indeed that's the case when we are checking applicability facets, but for invali facets in the requirements set what we get is invalid spec -> verification failure, which would then open an argument across the stakeholders

**Comment by aothms** [comment]
Well, it's a matter of how you look at things. If you're in a statically typed world, sure requesting an attribute that doesn't exist or comparing to a non-existent type is an error. But in a more data-oriented or open-world or flexible typed environment, a non-existent attribute would be null and types are just strings.

So I'm not necessarily convinced this is going to result in (heated) arguments.

It's also a spectrum:

- Invalid XML syntax
- Violation of XSD
- Invalid references to IFC constructs
- Contradictory expressions `(e.g applicability(entity=IfcWall) -> requirement(entity=IfcSlab)`

We will never get fully comprehensive towards the lower levels of this list. So this kind of dialogue between stakeholders will naturally also exist and will be good to shape future versions of IDS.

**Comment by CBenghi** [comment]
I agree; indeed my first post in this issue intended to start the conversation on what can be considered valid.

However at basic level, we could just enforce the agreements that have already been stated in the documentation (e.g. inverse attributes), and these should be fairly easy to implement in v1.0

Future releases will evolve as the implementers agree on reasonable levels of support.

**Comment by berlotti** [comment]
Anyone interested in building an IDS validator tool?
@CBenghi or @Moult  maybe?

**Comment by CBenghi** [comment]
I would. 
It would be probably be a c# .NET library coupled with a command line console app.
Would an MIT License do?

**Comment by Moult** [comment]
Sure, given there is an overlap with the existing IFC generated testcases, it seems a natural extension to build a validation library too. If we cover Python and @CBenghi covers C# I think that covers a lot of the stacks out there, the only stack that is a bit inconvenient is web I guess, but maybe the IFC.JS folks might join the fun later. In our case it would also be a lib + CLI app, but we might also extend the doc generator to make a "validator spec generator".

Is this something that should occur before the release of IDS v1, or should it happen in its own time?

**Comment by CBenghi** [comment]
I think it should happen before the release.

**Comment by pasi-paasiala** [comment]
The audit tool does this: https://github.com/buildingSMART/IDS-Audit-tool.
